tests/sync/session/session_util.hpp

```diff
@@ -0,0 +1,99 @@
+bool sessions_are_inactive(const SyncSession& session);
+
+template <typename... S>
+bool sessions_are_active(const SyncSession& session, const S&... s)
+{
+    return sessions_are_active(session) && sessions_are_active(s...);
+}
+
+template <typename... S>
+bool sessions_are_inactive(const SyncSession& session, const S&... s)
+{
+    return sessions_are_inactive(session) && sessions_are_inactive(s...);
+}
+
+}
+
+// Identical to `sync_session(...)`, but takes a bind-session callback that is
+// passed directly into the configuration. This allows, for example, a session
+// that remains stalled in the 'waiting for token' state.
+template <typename BindCallback, typename ErrorHandler>
+std::shared_ptr<SyncSession> full_control_sync_session(SyncServer& server, std::shared_ptr<SyncUser> user, const std::string& path,
+                                                       BindCallback&& bind_callback, ErrorHandler&& error_handler,
+                                                       SyncSessionStopPolicy stop_policy=SyncSessionStopPolicy::AfterChangesUploaded,
+                                                       std::string* on_disk_path=nullptr,
+                                                       util::Optional<Schema> schema=none,
+                                                       Realm::Config* out_config=nullptr)
+{
+    std::string url = server.base_url() + path;
+    SyncTestFile config({user, url, std::move(stop_policy),
+        std::forward<BindCallback>(bind_callback), std::forward<ErrorHandler>(error_handler)});
+    if (schema) {
+        config.schema = *schema;
+    }
+    if (on_disk_path) {
+        *on_disk_path = config.path;
+    }
+    if (out_config) {
+        *out_config = config;
+    }
+    std::shared_ptr<SyncSession> session;
+    {
+        auto realm = Realm::get_shared_realm(config);
+        session = SyncManager::shared().get_session(config.path, *config.sync_config);
+    }
+    return session;
+}
+
+// Convenience function for creating and configuring sync sessions for test use.
+// Many of the optional arguments can be used to pass information about the
+// session back out to the test, or configure the session more precisely.
+template <typename FetchAccessToken, typename ErrorHandler>
+std::shared_ptr<SyncSession> sync_session(SyncServer& server, std::shared_ptr<SyncUser> user, const std::string& path,
+                                          FetchAccessToken&& fetch_access_token, ErrorHandler&& error_handler,
+                                          SyncSessionStopPolicy stop_policy=SyncSessionStopPolicy::AfterChangesUploaded,
+                                          std::string* on_disk_path=nullptr,
+                                          util::Optional<Schema> schema=none,
+                                          Realm::Config* out_config=nullptr)
+{
+    return full_control_sync_session(server, std::forward<std::shared_ptr<SyncUser>>(user), path,
```

Comment:
More specifically, `std::forward` is only meaningful when used with [forwarding references](http://en.cppreference.com/w/cpp/language/reference#Forwarding_references), which are a subset of rvalue references.